include/jwt-cpp/base.h

```diff
@@ -108,39 +140,44 @@ namespace jwt {
 				size_t length = 0;
 
 				padding() = default;
-				padding(size_t count, size_t length) : count(count), length(length) {}
 
-				padding operator+(const padding& p) { return padding(count + p.count, length + p.length); }
+				padding(size_t c, size_t l) : count(c), length(l) {}
 
-				friend bool operator==(const padding& lhs, const padding& rhs) {
-					return lhs.count == rhs.count && lhs.length == rhs.length;
-				}
+				padding operator+(const padding& p) const { return padding{count + p.count, length + p.length}; }
 			};
 
-			inline padding count_padding(const std::string& base, const std::vector<std::string>& fills) {
-				for (const auto& fill : fills) {
-					if (base.size() < fill.size()) continue;
-					// Does the end of the input exactly match the fill pattern?
-					if (base.substr(base.size() - fill.size()) == fill) {
-						return padding{1, fill.length()} +
-							   count_padding(base.substr(0, base.size() - fill.size()), fills);
+			inline std::size_t string_len(string_view str) { return str.size(); }
+
+			template<class str_input_it>
+			padding count_padding(string_view base, str_input_it fillStart, str_input_it fillEnd) {
+				for (str_input_it fillIt = fillStart; fillIt != fillEnd; ++fillIt) {
+					std::size_t fillLen = string_len(*fillIt);
+					if (base.size() >= fillLen) {
+						std::size_t deltaLen = base.size() - fillLen;
+						// Does the end of the input exactly match the fill pattern?
+						if (base.substr(deltaLen) == *fillIt) {
+							return padding{1UL, fillLen} + count_padding(base.substr(0, deltaLen), fillStart, fillEnd);
+						}
 					}
 				}
 
 				return {};
 			}
 
-			inline std::string encode(const std::string& bin, const std::array<char, 64>& alphabet,
-									  const std::string& fill) {
+			inline std::string encode(string_view bin, const char* alphabet, string_view fill) {
 				size_t size = bin.size();
 				std::string res;
 
+				res.reserve((4UL * size) / 3UL);
```

Comment:
I am not referring to a failure in the sense that it creates an error, however it will not prevent multiple allocations. The formula assumes fill is a single character and will allocate accordingly. reserve however does not follow the usual growth factor, it will allocate exactly the capacity passed to it (assuming its larger than the current size) meaning if we later add the fill and its not a single char the string has to reallocate to fit the extra chars. What I am proposing is to add a small extra of 2*fill.size() so that the reserved size is guaranteed to be larger than everything it will get.